hasktorch/src/Torch/Vision/Darknet/Forward.hs

```diff
@@ -0,0 +1,530 @@
+  in loss_x + loss_y + loss_w + loss_h + loss_conf + loss_cls
+
+
+data YoloOutput
+  = YoloOutput
+      { x :: Tensor,
+        y :: Tensor,
+        w :: Tensor,
+        h :: Tensor,
+        predBoxes :: Tensor,
+        predConf :: Tensor,
+        predClass :: Tensor
+      }
+  deriving (Show)
+
+
+instance HasForward Yolo (Maybe Tensor, Tensor) Tensor where
+  forward yolo@Yolo {..} (train, input) =
+    let num_samples = D.size 0 input
+        grid_size = D.size 2 input
+        g = grid_size
+        num_anchors = length anchors
+        img_dim = shape input !! 2
+        stride = (fromIntegral img_dim) / (fromIntegral grid_size) :: Float
+        prediction = toPrediction yolo input
+        pred_boxes = toPredBox yolo prediction
+        (px,py,pw,ph) = pred_boxes
+        pred_cls = toPredClass prediction
+        pred_conf = toPredConf prediction
+     in case train of
+        Nothing -> ( D.cat
+               (D.Dim (-1))
+               [ stride `D.mulScalar` D.view [num_samples, -1, 4] (D.cat (D.Dim 3) [px,py,pw,ph]),
+                 D.view [num_samples, -1, 1] pred_conf,
+                 D.view [num_samples, -1, classes] pred_cls
+               ]
+             )
+        Just target ->
+          let ignore_thres = 0.5
+              build_target = toBuildTargets pred_boxes pred_cls target anchors ignore_thres
+          in totalLoss yolo prediction build_target
+
+
+data Layer
+  = LConvolution Convolution
+  | LConvolutionWithBatchNorm ConvolutionWithBatchNorm
+  | LMaxPool MaxPool
+  | LUpSample UpSample
+  | LRoute Route
+  | LShortCut ShortCut
+  | LYolo Yolo
+  deriving (Show, Generic, Parameterized)
+
+data Darknet = Darknet [(Index, Layer)] deriving (Show, Generic, Parameterized)
+
+instance Serialise Parameter where
+  encode p = encode p' where p' :: [Float] = asValue $ toDependent p
+  decode = IndependentTensor . (asTensor :: [Float] -> Tensor) <$> decode
+
+instance Serialise Tensor where
+  encode p = encode p' where p' :: [Float] = asValue $ p
+  decode = (asTensor :: [Float] -> Tensor) <$> decode
+
+loadFloats :: System.IO.Handle -> Tensor -> IO Tensor
+loadFloats handle tensor = do
+  let len = 4 * product (shape tensor)
+  v <- BS.hGet handle len
+  t <- D.clone tensor
+  D.withTensor t $ \ptr1 -> do
+    let (BSI.PS fptr _ len') = v
+    F.withForeignPtr fptr $ \ptr2 -> do
+      BSI.memcpy (F.castPtr ptr1) (F.castPtr ptr2) (min len len')
+      return t
+
+loadWeights :: Darknet -> String -> IO Darknet
+loadWeights (Darknet layers) weights_file = do
+  System.IO.withFile weights_file System.IO.ReadMode $ \handle -> do
+    _ <- BS.hGet handle (5 * 4) -- skip header
+    layers' <- forM layers $ \(i,layer) -> do
+      let cur_params = flattenParameters layer
+      new_params <- forM cur_params $ \param -> loadFloats handle (toDependent param) >>= makeIndependent
+      return $ (i,replaceParameters layer new_params)
+    return $ Darknet layers'
```

Comment:
Conv2d/BatchNorm's paramaters are loaded incorrectly.

Reply:
Any idea why? I would like to try to do some work on loading pytorch models in the near future as well.

Reply:
The data is stored in the weight file as follows.
1, Bias of batchnorm
2, Weight of batchnorm
3, RunningMean of batchnorm
4, RunningVar of batchnorm
5, Weight of Conv2d

Conv2d has no bias.
https://github.com/eriklindernoren/PyTorch-YOLOv3/blob/master/models.py#L282-L317

The list of flattenParameters includes the bias of Conv2d with batchnorm.
Conv2d without batchnorm has bias.
Loader customization is needed.
 

Reply:
It may be easier to fix the ADT.